src/core/transport/chttp2/stream_encoder.c

```diff
@@ -601,4 +608,15 @@ void grpc_chttp2_encode(grpc_stream_op *ops, size_t ops_count, int eof,
     begin_frame(&st, DATA);
   }
   finish_frame(&st, 1, eof);
+
+  if (need_unref) {
+    grpc_mdctx_lock(mdctx);
+    for (unref_op = 0; unref_op < curop; unref_op++) {
+      op = &ops[unref_op];
+      if (op->type != GRPC_OP_METADATA) continue;
+      if (!op->data.metadata) continue;
+      grpc_mdctx_locked_mdelem_unref(mdctx, op->data.metadata);
+    }
+    grpc_mdctx_unlock(mdctx);
+  }
```

Comment:
Overall, this is very nice, and will help reduce locks dramatically. Is there a potential tail latency issue under contention because of holding the lock for longer? I don't expect it (I would actually expect an improvement because of a reduction in lock contention), but it is something that we ought to have a way to benchmark. Not now; just wanted to put my thought about it in writing.